hnn_core/extracellular.py

```diff
@@ -0,0 +1,587 @@
+        except IndexError:
+            raise IndexError(f'the data contain {len(self)} trials, the '
+                             f'indices provided are out of range: {trial_no}')
+        return ExtracellularArray(self.positions,
+                                  conductivity=self.conductivity,
+                                  method=self.method,
+                                  times=self.times,
+                                  voltages=return_data)
+
+    def __repr__(self):
+        class_name = self.__class__.__name__
+        msg = (f'{self.n_contacts} electrodes, '
+               f'conductivity={self.conductivity}, method={self.method}')
+        if len(self._data) > 0:
+            msg += f' | {len(self._data)} trials, {len(self.times)} times'
+        else:
+            msg += ' (no data recorded yet)'
+        return f'<{class_name} | {msg}>'
+
+    def __len__(self):
+        return len(self._data)  # length == number of trials
+
+    def copy(self):
+        """Return a copy of the ExtracellularArray instance
+
+        Returns
+        -------
+        array_copy : instance of ExtracellularArray
+            A copy of the array instance.
+        """
+        return deepcopy(self)
+
+    @property
+    def times(self):
+        return self._times
+
+    @property
+    def voltages(self):
+        return self._data
+
+    @property
+    def sfreq(self):
+        """Return the sampling rate of the extracellular data."""
+        if len(self.times) == 0:
+            return None
+        elif len(self.times) == 1:
+            raise RuntimeError('Sampling rate is not defined for one sample')
+
+        dT = np.diff(self.times)
+        Tsamp = np.median(dT)
+        if np.abs(dT.max() - Tsamp) > 1e-3 or np.abs(dT.min() - Tsamp) > 1e-3:
+            raise RuntimeError(
+                'Extracellular sampling times vary by more than 1 us. Check '
+                'times-attribute for errors.')
+
+        return 1000. / Tsamp  # times are in in ms
+
+    def _reset(self):
+        self._data = list()
+        self._times = list()
+
+    def get_data(self, return_times=False):
+        """Get extracellular electrode voltages as a Numpy array
+
+        The data are returned as as a Numpy array with exactly 3 dimensions:
+        (n_trials, n_channels, n_times), where ``n_channels`` are in the same
+        order as defined in the positions-argument.
+
+        Parameters
+        ----------
+        return_times : bool (default: False)
+            If True, also return the sample times.
+
+        Returns
+        -------
+        data : np.ndarray of shape (n_trials, n_channels, n_times)
+            The electrode voltages
+        times : np.ndarray of shape (n_times, )
+            The sampling times
+        """
+        if return_times:
+            return np.array(self._data), np.array(self._times)
+        else:
+            return np.array(self._data)
+
+    def smooth(self, window_len):
```

Comment:
Test to make sure smoothed waveforms are the same shape as the original data?

Reply:
Not sure how to test "sameness" here?

Reply:
Oh I simply meant number of samples, not the actual waveform shape. Realizing this is a moot point though since the plotting would break immediately if the number of samples didn't match `times`